esphome/components/as5600/__init__.py

```diff
@@ -0,0 +1,228 @@
+        range = config[CONF_END_POSITION] - config[CONF_START_POSITION]
+
+        # but need to account for start position being greater than end position
+        # where the range rolls back around the 0 position
+        if config[CONF_END_POSITION] < config[CONF_START_POSITION]:
+            range = RESOLUTION + config[CONF_END_POSITION] - config[CONF_START_POSITION]
+
+        try:
+            range_validator(range)
+            return config
+        except cv.Invalid as e:
+            raise cv.Invalid(
+                f"The range between start and end position is invalid. It was was {range} but {e.error_message}"
+            )
+
+    return validator
+
+
+CONFIG_SCHEMA = cv.All(
+    cv.Schema(
+        {
+            cv.GenerateID(): cv.declare_id(AS5600Component),
+            cv.Optional(CONF_DIR_PIN): pins.gpio_input_pin_schema,
+            cv.Optional(CONF_DIRECTION, default="CLOCKWISE"): cv.enum(
+                DIRECTION, upper=True
+            ),
+            cv.Optional(CONF_WATCHDOG, default=False): cv.boolean,
+            cv.Optional(CONF_POWER_MODE, default="NOMINAL"): cv.enum(
+                POWER_MODE, upper=True, space=""
+            ),
+            cv.Optional(CONF_HYSTERESIS, default="NONE"): cv.enum(
+                HYSTERESIS, upper=True, space=""
+            ),
+            cv.Optional(CONF_SLOW_FILTER, default="16X"): cv.enum(
+                SLOW_FILTER, upper=True, space=""
+            ),
+            cv.Optional(CONF_FAST_FILTER, default="NONE"): cv.enum(
+                FAST_FILTER, upper=True, space=""
+            ),
+            cv.Optional(CONF_START_POSITION, default=0): position(),
+            cv.Optional(CONF_END_POSITION): position(),
+            cv.Optional(CONF_RANGE): position_range(),
+        }
+    )
+    .extend(cv.COMPONENT_SCHEMA)
+    .extend(i2c.i2c_device_schema(0x36)),
+    # ensure end_position and range are mutually exclusive
+    cv.has_at_most_one_key(CONF_END_POSITION, CONF_RANGE),
+    has_valid_range_config(),
+)
+
+
+async def to_code(config):
+    var = cg.new_Pvariable(config[CONF_ID])
+    await cg.register_component(var, config)
+    await i2c.register_i2c_device(var, config)
+
+    cg.add(var.set_direction(config[CONF_DIRECTION]))
+    cg.add(var.set_watchdog(config[CONF_WATCHDOG]))
+    cg.add(var.set_power_mode(config[CONF_POWER_MODE]))
+    cg.add(var.set_hysteresis(config[CONF_HYSTERESIS]))
+    cg.add(var.set_slow_filter(config[CONF_SLOW_FILTER]))
+    cg.add(var.set_fast_filter(config[CONF_FAST_FILTER]))
+    cg.add(var.set_start_position(config[CONF_START_POSITION]))
+
+    if dir_pin_config := config.get(CONF_DIR_PIN):
+        pin = await cg.gpio_pin_expression(dir_pin_config)
+        cg.add(var.set_dir_pin(pin))
+
+    if end_position_config := config.get(CONF_END_POSITION):
+        cg.add(var.set_end_position(end_position_config))
+
+    if range_config := config.get(CONF_RANGE):
+        cg.add(var.set_range(range_config))
```

Comment:
These need to have `None` checks as they could be 0 integers which would be falsy and be skipped over.
```suggestion
    if (end_position_config := config.get(CONF_END_POSITION, None)) is not None:
        cg.add(var.set_end_position(end_position_config))

    if (range_config := config.get(CONF_RANGE, None)) is not None:
        cg.add(var.set_range(range_config))
```